adal/src/main/java/com/microsoft/aad/adal/HashMapExtensions.java

```diff
@@ -71,22 +71,28 @@ private HashMapExtensions() {
             while (parameterTokenizer.hasMoreTokens()) {
                 String pair = parameterTokenizer.nextToken();
                 String[] elements = pair.split("=");
+                String value = null;
+                String key = null;
 
                 if (elements.length == 2) {
-                    String key = null;
-                    String value = null;
                     try {
                         key = StringExtensions.urlFormDecode(elements[0].trim());
                         value = StringExtensions.urlFormDecode(elements[1].trim());
                     } catch (UnsupportedEncodingException e) {
                         Logger.i(TAG, ADALError.ENCODING_IS_NOT_SUPPORTED.getDescription(), e.getMessage(), null);
                     }
-
-                    if (!StringExtensions.isNullOrBlank(key)
-                            && !StringExtensions.isNullOrBlank(value)) {
-                        result.put(key, value);
+                } else if (elements.length == 1) {
+                    try {
+                        key = StringExtensions.urlFormDecode(elements[0].trim());
+                        value = "";
+                    } catch (UnsupportedEncodingException e) {
+                        Logger.i(TAG, ADALError.ENCODING_IS_NOT_SUPPORTED.getDescription(), e.getMessage(), null);
                     }
                 }
+
+                if (!StringExtensions.isNullOrBlank(key)) {
```

Comment:
If the above `UnsupportedEncodingException` is thrown, won't this try to `put(null, null)`? Would it make more sense to skip these values?

Reply:
`put(null, null)` would never happen with this check on Line 93. If UnsupportedEncodingException is thrown when decoding the value part, we will log the exception and if the key is not null or empty, we will put the key-value pair with empty value, which is requested in this PR.

Reply:
but put (key, null) can now happen which would have never happened before.

Reply:
Hi @addev-ashish I think this is by design. The purpose of this PR is to accept empty paras like CAs rather than throw AuthenticationException.

Reply:
Just confirming: this _is_ by design. I'm OK with `null` but it does mean that we need to be _very intentional_ when checking for the presence of keys by using `map.containsKey()` and not just using `map.get(key) != null`.

Preferentially speaking, would `""` be a better value? @addev-ashish @heidijinxujia